File: packages/xyd-opencli2go/src/runtime.ts
import type { OpencliSpecJson, XOpenApiSecurity } from '@xyd-js/opencli';

import { screamingSnakeCase } from './naming';

/** Static, generic HTTP runtime copied verbatim into every generated project. */
export function runtimeGo(): string {
  return `package runtime

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// Request is a generic API request assembled by a generated command handler.
type Request struct {
	Method  string
	Path    string
	Query   url.Values
	Headers http.Header
	Body    []byte
}

// Do executes the request against the configured base URL (applying auth) and
// prints the response body to stdout.
func Do(ctx context.Context, req Request) error {
	endpoint := strings.TrimRight(BaseURL(), "/") + req.Path
	if len(req.Query) > 0 {
		endpoint += "?" + req.Query.Encode()
	}

	var body io.Reader
	if len(req.Body) > 0 {
		body = bytes.NewReader(req.Body)
	}

	httpReq, err := http.NewRequestWithContext(ctx, req.Method, endpoint, body)
	if err != nil {
		return err
	}
	for key, values := range req.Headers {
		for _, value := range values {
			httpReq.Header.Add(key, value)
		}
	}
	if len(req.Body) > 0 && httpReq.Header.Get("Content-Type") == "" {
		httpReq.Header.Set("Content-Type", "application/json")
	}
	applyAuth(httpReq)

	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	out, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	fmt.Println(string(out))
	if resp.StatusCode >= 400 {
		return fmt.Errorf("request failed: %s", resp.Status)
	}
	return nil
}
`;
}

function authBlock(scheme: XOpenApiSecurity): string | null {
  const env = scheme.envVar || 'API_KEY';
  const name = scheme.name || '';
  switch (scheme.kind) {
    case 'bearer':
      return `	if v := os.Getenv(${JSON.stringify(env)}); v != "" {
		req.Header.Set("Authorization", "Bearer "+v)
	}`;
    case 'apiKey-header':
      return `	if v := os.Getenv(${JSON.stringify(env)}); v != "" {
		req.Header.Set(${JSON.stringify(name)}, v)
	}`;
    case 'apiKey-query':
      return `	if v := os.Getenv(${JSON.stringify(env)}); v != "" {
		q := req.URL.Query()
		q.Set(${JSON.stringify(name)}, v)
		req.URL.RawQuery = q.Encode()
	}`;
    case 'apiKey-cookie':
      return `	if v := os.Getenv(${JSON.stringify(env)}); v != "" {
		req.AddCookie(&http.Cookie{Name: ${JSON.stringify(name)}, Value: v})
	}`;
    case 'basic':
      return `	user := os.Getenv(${JSON.stringify(`${env}_USERNAME`)})
	pass := os.Getenv(${JSON.stringify(`${env}_PASSWORD`)})
	if user != "" || pass != "" {
		req.SetBasicAuth(user, pass)
	}`;
    default:
      return null;
  }
}

/** Generated config: the base URL + auth, derived from the spec's root x-openapi. */
export function configGo(spec: OpencliSpecJson, binName: string, baseURL: string): string {
  const prefix = screamingSnakeCase(binName);
  const security = spec['x-openapi']?.security || [];
  const blocks = security.map(authBlock).filter((b): b is string => b !== null);
  const authBody = blocks.length ? `\n${blocks.join('\n')}\n` : '';

  return `package runtime

import (
	"net/http"
	"os"
)

const defaultBaseURL = ${JSON.stringify(baseURL)}

// BaseURL returns the API base URL, overridable via ${prefix}_BASE_URL.
func BaseURL() string {
	if v := os.Getenv(${JSON.stringify(`${prefix}_BASE_URL`)}); v != "" {
		return v
	}
	return defaultBaseURL
}

// applyAuth attaches credentials read from the environment to the request.
func applyAuth(req *http.Request) {${authBody}}
`;
}

File: crates/xyd_opencli2go/src/runtime.rs
//! Runtime emission — port of runtime.ts. `runtime.go` is the fixed generic
//! HTTP runtime (vendored verbatim as an asset, exactly as the JS hardcodes
//! it); `config.go` is templated from the spec's root x-openapi security.

use serde_json::Value;

use crate::naming::screaming_snake_case;

/// Static, generic HTTP runtime copied verbatim into every generated project.
pub fn runtime_go() -> String {
    include_str!("runtime.go.txt").to_string()
}

fn q(s: &str) -> String {
    serde_json::to_string(s).expect("string serializes")
}

/// The auth block for one security scheme (None → unsupported kind, skipped).
fn auth_block(scheme: &Value) -> Option<String> {
    let env = scheme
        .get("envVar")
        .and_then(|v| v.as_str())
        .unwrap_or("API_KEY");
    let name = scheme.get("name").and_then(|v| v.as_str()).unwrap_or("");
    let kind = scheme.get("kind").and_then(|v| v.as_str()).unwrap_or("");
    match kind {
        "bearer" => Some(format!(
            "\tif v := os.Getenv({}); v != \"\" {{\n\t\treq.Header.Set(\"Authorization\", \"Bearer \"+v)\n\t}}",
            q(env)
        )),
        "apiKey-header" => Some(format!(
            "\tif v := os.Getenv({}); v != \"\" {{\n\t\treq.Header.Set({}, v)\n\t}}",
            q(env),
            q(name)
        )),
        "apiKey-query" => Some(format!(
            "\tif v := os.Getenv({}); v != \"\" {{\n\t\tq := req.URL.Query()\n\t\tq.Set({}, v)\n\t\treq.URL.RawQuery = q.Encode()\n\t}}",
            q(env),
            q(name)
        )),
        "apiKey-cookie" => Some(format!(
            "\tif v := os.Getenv({}); v != \"\" {{\n\t\treq.AddCookie(&http.Cookie{{Name: {}, Value: v}})\n\t}}",
            q(env),
            q(name)
        )),
        "basic" => Some(format!(
            "\tuser := os.Getenv({})\n\tpass := os.Getenv({})\n\tif user != \"\" || pass != \"\" {{\n\t\treq.SetBasicAuth(user, pass)\n\t}}",
            q(&format!("{env}_USERNAME")),
            q(&format!("{env}_PASSWORD"))
        )),
        _ => None,
    }
}

/// Generated config: base URL + auth, from the spec's root x-openapi.
pub fn config_go(spec: &Value, bin_name: &str, base_url: &str) -> String {
    let prefix = screaming_snake_case(bin_name);
    let security = spec
        .get("x-openapi")
        .and_then(|x| x.get("security"))
        .and_then(|s| s.as_array())
        .cloned()
        .unwrap_or_default();
    let blocks: Vec<String> = security.iter().filter_map(auth_block).collect();
    let auth_body = if blocks.is_empty() {
        String::new()
    } else {
        format!("\n{}\n", blocks.join("\n"))
    };

    format!(
        "package runtime\n\nimport (\n\t\"net/http\"\n\t\"os\"\n)\n\nconst defaultBaseURL = {}\n\n// BaseURL returns the API base URL, overridable via {prefix}_BASE_URL.\nfunc BaseURL() string {{\n\tif v := os.Getenv({}); v != \"\" {{\n\t\treturn v\n\t}}\n\treturn defaultBaseURL\n}}\n\n// applyAuth attaches credentials read from the environment to the request.\nfunc applyAuth(req *http.Request) {{{auth_body}}}\n",
        q(base_url),
        q(&format!("{prefix}_BASE_URL"))
    )
}
